magenta/music/musicxml_parser.py

```diff
@@ -0,0 +1,678 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""MusicXML parser.
+
+Simple MusicXML parser used to convert MusicXML
+into tensorflow.magenta.NoteSequence.
+"""
+
+# Imports
+# Python 2 uses integer division for integers. Using this gives the Python 3
+# behavior of producing a float when dividing integers
+from __future__ import division
+
+import xml.etree.ElementTree as ET
+from fractions import Fraction
+from zipfile import ZipFile
+from magenta.music import constants
+
+
+class MusicXMLParserState(object):
+  """Maintains internal state of the MusicXML parser"""
+  def __init__(self):
+    # Default to one division per measure
+    self.divisions = 1
+
+    # Default to a tempo of 120 beats per minute
+    # (This is the default tempo according to the
+    # Standard MIDI Files 1.0 Specification)
+    self.tempo = 120
+
+    # Duration of a single beat in seconds
+    self.seconds_per_beat = 0.5
+
+    # Running total of time for the current event.
+    # Resets to 0 on every part. Affected by <forward> and <backup> elements
+    self.time_position = 0
+
+    # Default to a MIDI velocity of 64 (mf)
+    self.velocity = 64
+
+    # Default MIDI program (1 = grand piano)
+    self.midi_program = 1
+
+    # Current MIDI channel (usually equal to the part number)
+    self.midi_channel = 1
+
+    # Keep track of previous note to get chord timing correct
+    self.previous_note = None
+
+    # Keep track of current transposition level
```

Comment:
Yes, it can be positive or negative semitones, so I updated the comment. Note that it does not support quarter tones or tuning systems other than equal temperament.